### PR TITLE
Provide semantic functionality for packages that are subdirectories of a workspace folder

### PR DESCRIPTION
**Pull request by ahoppen**
rdar://119740711

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/LSPLogging/CustomLogStringConvertible.swift [hunk 1]
```diff
@@ -67,3 +67,21 @@ extension String {
     return Insecure.MD5.hash(data: Data(self.utf8)).description
   }
 }
+
+private struct OptionalWrapper<Wrapped>: CustomLogStringConvertible where Wrapped: CustomLogStringConvertible {
```
Comment (bnbarham) [inline Sources/LSPLogging/CustomLogStringConvertible.swift:71]: Were optionals just redacted previously 😅?

**Review by bnbarham — APPROVED** [review]

File: Sources/SourceKitLSP/SourceKitServer.swift [hunk 2]
```diff
@@ -407,20 +408,24 @@ public actor SourceKitServer {
...
-    set {
-      self.workspaces = newValue
-    }
+  var workspaces: [Workspace] {
+    return workspacesAndIsImplicit.map(\.workspace)
+  }
+
+  @_spi(Testing)
+  public func setWorkspaces(_ newValue: [(workspace: Workspace, isImplicit: Bool)]) {
+    self.workspacesAndIsImplicit = newValue
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SourceKitLSP/SourceKitServer.swift:428]: The problem is that `SourceKitServer` is an actor and you can’t await the setting of a property. So you need a setter.

File: Sources/SourceKitLSP/SourceKitServer.swift [hunk 3]
```diff
@@ -1447,18 +1514,20 @@ extension SourceKitServer {
       preChangeWorkspaces[docUri] = await self.workspaceForDocument(uri: docUri)
     }
     if let removed = notification.event.removed {
-      self.workspaces.removeAll { workspace in
-        return removed.contains(where: { workspaceFolder in
-          workspace.rootUri == workspaceFolder.uri
-        })
+      self.workspacesAndIsImplicit.removeAll { workspace in
+        // Close all implicit workspaces as well because
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SourceKitLSP/SourceKitServer.swift:1518]: Oh, whoops

File: Sources/LSPLogging/CustomLogStringConvertible.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/LSPLogging/CustomLogStringConvertible.swift:71]: We never tried to log an optional `CustomLogStringConvertible` before, so the question just didn’t come up.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**